### PR TITLE
Add nonce support.

### PR DESCRIPTION
**Pull request by arturkow2000**

### CODE REVIEW

File: src/tpm2-crypto.c [hunk 1]
```diff
@@ -529,13 +529,26 @@ static UsefulBuf concat_data_sign(UsefulBuf data, UsefulBuf sign, UsefulBuf buf)
     }
 }
 
-UsefulBuf sign_with_aik(UsefulBuf data)
+UsefulBuf sign_with_aik(UsefulBuf data, UsefulBuf nonce)
```
Comment (krystian-hebel) [inline src/tpm2-crypto.c]:
```suggestion
UsefulBuf sign_with_aik(UsefulBuf data, UsefulBufC nonce)
```
Plus drop `UsefulBuf_Unconst()` from calls to this function.

File: src/tpm2-crypto.c [hunk 2]
```diff
@@ -808,7 +821,7 @@ static UsefulBuf get_pcr_assertions(uint32_t pcrs)
     return ret;
 }
 
-UsefulBuf get_signed_rim(uint32_t pcrs)
+UsefulBuf get_signed_rim(uint32_t pcrs, UsefulBuf nonce)
```
Comment (krystian-hebel) [inline src/tpm2-crypto.c]:
```suggestion
UsefulBuf get_signed_rim(uint32_t pcrs, UsefulBufC nonce)
```

File: src/tpm2-crypto.c [hunk 3]
```diff
@@ -529,13 +529,26 @@ static UsefulBuf concat_data_sign(UsefulBuf data, UsefulBuf sign, UsefulBuf buf)
...
+        memcpy(data_and_nonce.ptr, data.ptr, data.len);
+        memcpy((uint8_t*)data_and_nonce.ptr + data.len, nonce.ptr, nonce.len);
+    }
+
+    if (UsefulBuf_IsNULLOrEmpty(data_and_nonce)) {
+        signature = get_aik_signature(data);
+    } else {
+        signature = get_aik_signature(data_and_nonce);
+        free(data_and_nonce.ptr);
+    }
```
Comment (krystian-hebel) [inline src/tpm2-crypto.c]:
```suggestion
    if (!UsefulBuf_IsNULLOrEmpty(nonce)) {
        data_and_nonce.len = data.len + nonce.len;
        data_and_nonce.ptr = malloc(data_and_nonce.len);
        memcpy(data_and_nonce.ptr, data.ptr, data.len);
        memcpy((uint8_t*)data_and_nonce.ptr + data.len, nonce.ptr, nonce.len);

        signature = get_aik_signature(data_and_nonce);
        free(data_and_nonce.ptr);
    } else {
        signature = get_aik_signature(data);
    }
```
`data_and_nonce` can also be limited to this block.

File: src/fobnail-attester.c [hunk 4]
```diff
@@ -20,6 +21,31 @@ static volatile sig_atomic_t quit = 0;
...
+    QCBORDecodeContext ctx;
+    UsefulBufC nonce;
+
+    if (UsefulBuf_IsNULLOrEmptyC(data))
+        return NULLUsefulBufC;
+
+    QCBORDecode_Init(&ctx, data, QCBOR_DECODE_MODE_NORMAL);
+    QCBORDecode_EnterMap(&ctx, NULL);
+        QCBORDecode_GetByteStringInMapSZ(&ctx, "nonce", &nonce);
+    QCBORDecode_ExitMap(&ctx);
```
Comment (krystian-hebel) [inline src/fobnail-attester.c:39]: @arturkow2000 have you tested if this works with CBOR maps that also have other fields than just `nonce`?

File: src/fobnail-attester.c [hunk 5]
```diff
@@ -189,9 +215,15 @@ static void coap_metadata_handler(struct coap_resource_t* resource, struct coap_
     int ret;
     UsefulBuf ub_meta;
     UsefulBuf ub;
+    UsefulBufC request;
+    UsefulBufC nonce = NULLUsefulBufC;
     struct meta_data meta;
     printf("Received message: %s\n", coap_get_uri_path(in)->s);
 
+    if (coap_get_data(in, &request.len, (const uint8_t**)&request.ptr)) {
```
Comment (krystian-hebel) [inline src/fobnail-attester.c]: I'd rather see `get_nonce` taking `in` as a parameter and calling this function internally, so it won't have to be repeated for every URI that uses nonce.

**Review by krystian-hebel — CHANGES_REQUESTED** [review]

File: src/tpm2-crypto.c (hunk 1 above)
Comment (arturkow2000) [reply src/tpm2-crypto.c]: Fixed

File: src/fobnail-attester.c (hunk 4 above)
Comment (arturkow2000) [reply src/fobnail-attester.c:39]: Just tested by adding few fields, still works.

File: src/fobnail-attester.c (hunk 5 above)
Comment (arturkow2000) [reply src/fobnail-attester.c]: Done

**Review by krystian-hebel — APPROVED** [review]

**Pull request merged**